### PR TITLE
Add test for scenario that polymorphic association type is nil

### PR DESCRIPTION
**Pull request by wendy-clio**
This is a follow-up to the [slack question](https://clio.slack.com/archives/C60TARCAF/p1714168920731479). The spec proves that,
1. With mixed polymorphic association records that one has `type = nil`, we still load the rest of the valid association.
2. The publish the `N+1` notification for the record that has `type = nil`

We will have some other future PR to fix the problem of this false alarm `N+1` notification

### CODE REVIEW

File: spec/lib/jit_preloader/preloader_spec.rb [hunk 1]
```diff
@@ -165,6 +165,33 @@
...
       end
+
+      context "when a record has a polymorphic association type is nil" do
+        before do
+          contact1.update!(contact_owner_type: nil, contact_owner_id: nil)
+        end
+
+        it "successfully load the rest of association values" do
+          contacts = Contact.jit_preload.to_a
+          expect(contacts.first.contact_owner).to eq(nil)
```
Comment (apalmblad) [inline spec/lib/jit_preloader/preloader_spec.rb:176]: So the query run here - is it a false alarm or is actually running a query?  Is there  anything we can do to mark the association as loaded isntead of running a query?

Comment (wendy-clio) [reply spec/lib/jit_preloader/preloader_spec.rb:176]: It's running the preload but ActiveRecord marks the preload as `loaded unsuccessfully` since it can not successfully interpret the return type (which is nil in this case).

My primitive idea is to add another check on here https://github.com/clio/jit_preloader/blob/b1a60fde8d4cd3054735324bc3349d50395b1377/lib/jit_preloader/active_record/associations/singular_association.rb#L5
that just return if we found a record match the following conditions
1. the record target association to load is a polymorphic one, this should easily check on the reflection itself.
2. the record association type is nil.

But that will have to come as another PR.

File: spec/lib/jit_preloader/preloader_spec.rb [hunk 2]
```diff
@@ -165,6 +165,33 @@
...
+          expect(contacts.third.contact_owner).to eq(Address.first)
+        end
+
+        it "publish N+1 notification due to polymorphic nil type" do
+          contacts = Contact.jit_preload.to_a
+
+          ActiveSupport::Notifications.subscribed(callback, "n_plus_one_query") do
+            contacts.first.contact_owner
+          end
+        end
```
[the remark this answers is not in the document]
Comment (wendy-clio) [reply spec/lib/jit_preloader/preloader_spec.rb]: Thanks for the catch, indeed I miss a expect. Refactored!

File: spec/lib/jit_preloader/preloader_spec.rb (hunk 1 above)
Comment (wendy-clio) [reply spec/lib/jit_preloader/preloader_spec.rb:176]: no, cause running the `contacts.first.contact_owner` is the first time it preload the `contact_owner` association which should execute the query. The following access for the same association is the part it should not execute another query to access the data.

**Review by tudorbertiean — DISMISSED** [review]
LGTM - happy to have this in for now so that we can have the confidence + history to remove the n+1 spec once we have the fix

File: spec/lib/jit_preloader/preloader_spec.rb (hunk 1 above)
Comment (apalmblad) [reply spec/lib/jit_preloader/preloader_spec.rb:176]: If you repeat this line in the `expect { ... }.not_to make_dsatabase_queries` section what happens?  That seems like that would prove the issue is a false positive; at the moment, we've asserted that for two of the records we don't have a problem with our polymorphic code, why not the first?

Comment (wendy-clio) [reply spec/lib/jit_preloader/preloader_spec.rb:176]: This is my local outcome if we wrap the first one in expect too
image

with code
```
it "successfully load the rest of association values" do
            contacts = Contact.jit_preload.to_a
            expect do
              contacts.first.contact_owner
            end.not_to make_database_queries
  
            expect do
              contacts.second.contact_owner
              contacts.third.contact_owner
            end.not_to make_database_queries
  
            expect(contacts.second.contact_owner).to eq(ContactOwner.first)
            expect(contacts.third.contact_owner).to eq(Address.first)
end
```

Comment (wendy-clio) [reply spec/lib/jit_preloader/preloader_spec.rb:176]: sure, I can do that.

**Review by MrJakeReid — APPROVED** [review]
We walked through this at standup this morning, and I think the test expansion for asserting when the `contacts.first` call does/does not make an extra database call makes me more confident this was a transient.

**Review by apalmblad — APPROVED** [review]
